sdk/storage/azure-storage-common/inc/azure/storage/common/paged_response.hpp

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <type_traits>
+
+#include <azure/core/context.hpp>
+#include <azure/core/http/raw_response.hpp>
+
+namespace Azure { namespace Storage {
+
+  namespace _internal {
+    template <class Derived> class PageResult {
+    public:
+      std::unique_ptr<Azure::Core::Http::RawResponse> RawResponse;
+      std::string NextPageToken;
+      std::string CurrentPageToken;
+
+      bool HasMore() const { return !NextPageToken.empty(); }
+
+      void NextPage(const Azure::Core::Context& context)
+      {
+        if (!HasMore())
+        {
+          // NextPage() should NEVER be called if HasMore() is false
+          std::abort(); // or throw std::out_of_range
+        }
+
+        static_assert(
+            std::is_base_of<PageResult, Derived>::value,
+            "The template argument \"Derived\" should derive from PageResult<Derived>.");
+
+        CurrentPageToken = NextPageToken;
+        static_cast<Derived*>(this)->OnNextPage(context);
+      }
+
+    protected:
+      PageResult() = default;
+      PageResult(PageResult&&) = default;
+      PageResult& operator=(PageResult&&) = default;
+    };
+
+  } // namespace _internal
+
+  template <class T> class Pageable {
+  public:
+    explicit Pageable(T value) : m_value(std::move(value))
+    {
+      static_assert(
+          std::is_base_of<_internal::PageResult<T>, T>::value,
+          "The template argument should derive from PageResult");
+    }
+
+    class Iterator {
```

Comment:
1) Does the Iterator class have to be public?
2) If we decide to not make Pageable<T> today, can we make it iteratable in the future by adding this back, in a non-breaking way?

If so, maybe we can leave out iterator for now and revisit it later once we get some usability feedback from customers that request it.

Reply:
1. Does the Iterator class have to be public?

I think so.

2. If we decide to not make Pageable today, can we make it iteratable in the future by adding this back, in a non-breaking way?

If we decide to not make Pageable today, I guess we'll never have it. It'll be a breaking change.

Reply:
> If we decide to not make Pageable today, can we make it iteratable in the future by adding this back, in a non-breaking way?

Sorry, I meant if we decide to not make `Pageable<T>` **iteratable** today, can we make it iteratable in the future?

Reply:
> Sorry, I meant if we decide to not make Pageable<T> iteratable today, can we make it iteratable in the future?

I don't think so. To make iterator to work, I have to add the Pageable<T> class. The return type is different from the first prototype by Jeffrey. It's like a breaking  change.